reporter/src/main/kotlin/reporters/evaluatedmodel/EvaluatedModel.kt

```diff
@@ -148,7 +148,8 @@ data class EvaluatedModel(
             YAMLMapper().apply(mapperConfig).registerModule(IntIdModule(INT_ID_TYPES))
         }
 
-        fun create(input: ReporterInput): EvaluatedModel = EvaluatedModelMapper(input).build()
+        fun create(input: ReporterInput, deduplicateDependencyTree: Boolean = false): EvaluatedModel =
```

Comment:
Can you help me understanding this change with the following questions?

1. Does it prune redundant subtrees if they are from same scope?
2. Does it prune redundant subtrees if they are in different scope?
3. Is the logical tree affected or is it just the representation?
4. How fast is this code to execute?

Reply:
Hi Frank,

ad 1: Yes, redundant subtrees are pruned from the same scope.
ad 2: No, subtrees are not pruned over different scopes. So if a subtree appears multiple times in different scopes, each scope includes one copy.
ad 3: I am not exactly sure what you mean by "logical tree". Nodes are actually removed from the tree to reduce its size. Users must be aware that they do not see all levels of the tree in the WebApp report when they set this option.
ad 4: With this option enabled, the report generation may even run faster, since only parts of the dependency graph are traversed. For the projects that caused us problems, the reporter step failed after 30 to 40 minutes. With the option it is successful after < 3 minutes.

Reply:
To add one example where this change is required: We saw one Android project with many dependencies on Android core libraries where each spans a very large and highly redundant dependency tree. IIRC the largest scope had a dependency tree with more than 300,000 nodes, and over all scopes there were about 10,000,000 nodes. So even if the reporter would not eventually run out of memory, it would be impossible to load the generated web app report in any browser and render the dependency tree.

Reply:
Thanks @oheger-bosch for the clarifications!

With #3 I intended to ask whether it would make sense to actually do the propose change to the evaluated model while making the webapp still show the full tree. What do you think?

Reply:
> while making the webapp still show the full tree.

Didn't @mnonnenmacher basically [already answer that](https://github.com/oss-review-toolkit/ort/pull/5471#discussion_r903077866)?

> it would be impossible to load the generated web app report in any browser and render the dependency tree.

I'm not sure how "smart" the rendering in the web app reporter is, though. Could be that collapsed parts of the tree / graph do not really take up much memory, and / or that duplicate subtrees could be rendered as "references" instead of "copies".

Reply:
So you mean, the JavaScript code would be modified, so that it can render the full tree even if the model stores a compressed representation? This would be a good solution; however, I have no clue how much effort this would mean.

I had the idea to use a representation similar to the dependency graph. Maybe there are JavaScript libraries available that can render a graph with such a representation. This would probably look rather fancy. But I am no JavaScript expert and do not know which libraries could be used for this purpose.

Reply:
> Maybe there are JavaScript libraries available that can render a graph with such a representation.

In this context it could make sense to look at https://github.com/snyk/dep-graph again. I don't see any rendering happening in there, though, but maybe there's a companion library to do so.

Reply:
> Didn't @mnonnenmacher basically [already answer that](https://github.com/oss-review-toolkit/ort/pull/5471#discussion_r903077866)?

@sschuberth - No, I was referring to what @oheger-bosch  writes below, e.g. "the JavaScript code would be modified, so that it can render the full tree even if the model stores a compressed representation"